storage/src/mmr/mutable.rs

```diff
@@ -140,61 +162,78 @@ impl<K: Array, V: Array, H: CHasher> MutableMmr<K, V, H> {
 
     /// Get the number of operations that have been applied to this store.
     pub fn op_count(&self) -> u64 {
-        self.log.len() as u64
+        self.log.len() as u64 + self.pruned_loc
     }
 
     /// Updates `key` to have value `value`.  If the key already has this same value, then this is a
     /// no-op.
+    ///
+    /// Also move exactly one active operation to tip if at least one is movable.
     pub fn update(&mut self, hasher: &mut H, key: K, value: V) {
-        let new_loc = self.log.len();
+        let new_loc = self.log.len() as u64 + self.pruned_loc;
 
         // Update the snapshot.
         if let Some(loc) = self.snapshot.get_mut(&key) {
-            let last_value = match self.log[*loc].op_type {
+            let i = *loc - self.pruned_loc;
+            let last_value = match self.log[i as usize].op_type {
                 Type::Deleted => panic!("deleted key should not be in snapshot: {}", key),
                 Type::Update(ref v) => v,
             };
             if value == *last_value {
                 // Trying to assign the same value is a no-op.
                 return;
             }
+            if *loc == self.inactivity_floor_loc {
+                self.inactivity_floor_loc += 1;
+            }
             *loc = new_loc;
         } else {
             self.snapshot.insert(key.clone(), new_loc);
         }
 
         let op = Type::Update(value.clone());
         self.apply_op(hasher, key, op);
-    }
-
-    /// Update the log and operations MMR with the given key and operation.
-    fn apply_op(&mut self, hasher: &mut H, key: K, op_type: Type<V>) {
-        let op = Operation { key, op_type };
 
-        // Update the ops MMR.
-        let digest = Self::op_digest(hasher, &op);
-        self.ops.add(hasher, &digest);
-
-        // Append the operation to the log.
-        self.log.push(op);
+        // Move exactly one active operation to tip if at least one is movable.
+        self.raise_inactivity_floor(hasher, 1);
     }
 
     /// Delete `key` and its value from the store. Deleting a key that already has no value is a
     /// no-op.
+    ///
+    /// Also move exactly one active operation to tip if at least one is movable.
     pub fn delete(&mut self, hasher: &mut H, key: K) {
         // Remove the key from the snapshot.
-        if self.snapshot.remove(&key).is_none() {
+        let Some(loc) = self.snapshot.remove(&key) else {
             return;
+        };
+        if loc == self.inactivity_floor_loc {
+            self.inactivity_floor_loc += 1;
         }
 
         self.apply_op(hasher, key, Type::Deleted);
+
+        // Move exactly one active operation to tip if at least one is movable.
+        self.raise_inactivity_floor(hasher, 1);
     }
 
     /// Return the root hash of the mutable MMR.
     pub fn root(&self, hasher: &mut H) -> H::Digest {
```

Comment:
I suspect we'll want to hash this with the inactivity floor (so folks syncing know where to start). I'm not sure if we'd want to include that in the block or deal with it here. I suppose just encoding it in the block should be sufficient (and less opinionated).

Reply:
The inactivity floor should be hidden from the user IMO but perhaps you meant the "prune_loc", which influences what remains provable.  Still feels awkward to me to reflect that in the root hash but I'll have to think about that a bit more.